plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -59,18 +64,27 @@ public class QueryInsightsService extends AbstractLifecycleComponent {
     /**
      * Constructor of the QueryInsightsService
      *
-     * @param threadPool     The OpenSearch thread pool to run async tasks
+     * @param clusterSettings OpenSearch cluster level settings
+     * @param threadPool The OpenSearch thread pool to run async tasks
+     * @param client OS client
      */
     @Inject
-    public QueryInsightsService(final ThreadPool threadPool) {
+    public QueryInsightsService(final ClusterSettings clusterSettings, final ThreadPool threadPool, final Client client) {
         enableCollect = new HashMap<>();
         queryRecordsQueue = new LinkedBlockingQueue<>(QueryInsightsSettings.QUERY_RECORD_QUEUE_CAPACITY);
+        this.threadPool = threadPool;
+        final QueryInsightsExporterFactory queryInsightsExporterFactory = new QueryInsightsExporterFactory(client);
+        // initialize top n queries services and configurations consumers
         topQueriesServices = new HashMap<>();
         for (MetricType metricType : MetricType.allMetricTypes()) {
             enableCollect.put(metricType, false);
-            topQueriesServices.put(metricType, new TopQueriesService(metricType));
+            topQueriesServices.put(metricType, new TopQueriesService(metricType, threadPool, queryInsightsExporterFactory));
```

Comment:
We should not provide topQueriesService to create arbitrary exporter objects using the factory. Also, the factory should keep track of exporter objects being created

Reply:
> We should not provide topQueriesService to create arbitrary exporter objects using the factory

Make sense! does it make sense to have a map to record the "service to allowed exporters list" mapping in the factory? Also since now we only have 2 exporters and top queries services should be able to access both of them. We can do this as a follow up once we add more types of exporters?

> the factory should keep track of exporter objects being created

This is a very good point. We should keep track of all exporters created and close them when the service is stopped.